core/src/main/scala/org/bykn/bosatsu/codegen/clang/ClangGen.scala

```diff
@@ -1168,35 +1169,43 @@ object ClangGen {
               Eval.now(Right((s, a)))
             )
 
+          def update[A](fn: State => (State, A)): T[A] =
+            StateT(s => EitherT[Eval, Error, (State, A)](Eval.now(Right(fn(s)))))
+
+          def tryUpdate[A](fn: State => Either[Error, (State, A)]): T[A] =
+            StateT(s => EitherT[Eval, Error, (State, A)](Eval.now(fn(s))))
+
+          def tryRead[A](fn: State => Either[Error, A]): T[A] =
+            StateT(s => EitherT[Eval, Error, (State, A)](Eval.now(fn(s).map((s, _)))))
+
           def globalIdent(pn: PackageName, bn: Bindable): T[Code.Ident] =
-            StateT { s =>
+            tryUpdate { s =>
               s.externals(pn, bn) match {
                 case Some((incl, ident, _)) =>
                   // TODO: suspect that we are ignoring arity here
                   val withIncl = s.include(incl)
-                  result(withIncl, ident)
+                  Right((withIncl, ident))
                 case None =>
                   val key = (pn, bn)
                   s.allValues.get(key) match {
-                    case Some((_, ident)) => result(s, ident)
-                    case None => errorRes(Error.UnknownValue(pn, bn))
+                    case Some((_, ident)) => Right((s, ident))
+                    case None => Left(Error.UnknownValue(pn, bn))
                   }
               }
             }
 
           def bind[A](bn: Bindable)(in: T[A]): T[A] = {
-            val init: T[Unit] = StateT { s =>
+            val init: T[Unit] = update { s =>
+              val cnt = s.bindCount
               val v = s.binds.get(bn) match {
-                case None => NonEmptyList.one(Right(0))
-                case Some(items @ NonEmptyList(Right(idx), _)) =>
-                  Right(idx + 1) :: items
-                case Some(items @ NonEmptyList(Left((_, idx)), _)) =>
-                  Right(idx + 1) :: items
+                case None => NonEmptyList.one(Right(cnt))
+                case Some(items) =>
+                  Right(cnt) :: items
               }  
-              result(s.copy(binds = s.binds.updated(bn, v)), ())
+              (s.copy(bindCount = cnt + 1, binds = s.binds.updated(bn, v)), ())
```

Comment:
this was a bug. We need each scope to be a distinct variable in C, but popping off the stack reused the variable names. With some of the inlining changes, we triggered the bug. The fix here is to just allocate a new ID for each binding inside of a top level name and reset on each value.